frontend/src/main/scala/bloop/logging/BspServerLogger.scala

```diff
@@ -133,15 +137,43 @@ final class BspServerLogger private (
       })
   }
 
+  private def toExistingBspPosition(pos: xsbti.Position): Option[bsp.Range] = {
+    def asIntPos(opt: Optional[Integer]) = InterfaceUtil.toOption(opt).map(_.toInt)
+    for {
+      startLine <- asIntPos(pos.startLine())
+      startColumn <- asIntPos(pos.startColumn())
+      endLine <- asIntPos(pos.endLine())
+      endColumn <- asIntPos(pos.endColumn())
+    } yield bsp.Range(
+      bsp.Position(startLine - 1, startColumn),
+      bsp.Position(endLine - 1, endColumn)
+    )
+  }
+
+  private def toScalaDiagnostic(actions: java.util.List[xsbti.Action]): bsp.ScalaDiagnostic = {
+    val bspActions = actions.asScala.map { action =>
+      val description = InterfaceUtil.toOption(action.description())
+      val edits = action.edit().changes().asScala.flatMap { edit =>
+        toExistingBspPosition(edit.position()).map { range =>
+          bsp.ScalaTextEdit(range, edit.newText())
+        }
+      }
+      val workspaceEdit = bsp.ScalaWorkspaceEdit(Option(edits.toList))
+      bsp.ScalaAction(action.title(), description, Some(workspaceEdit))
+    }
+    bsp.ScalaDiagnostic(Some(bspActions.toList))
```

Comment:
So for a lot of things these actions will be empty. I see in the tests we end up with a bunch of `"actions": []` and I'm curious if that's what we want? Do you think in the situations where we don't have any actions we should instead just return `None` instead of `Some(emptyList)`? I'm not sure if there is any difference here for this scenario or if it really matters. But maybe in situations where we're reporting a ton of diagnostics it'll keep the payload smaller to not have a bunch of empty actions? I'm not really sure if it matters or not.

Reply:
The sbt interfaces do not have actions as optional, so I assumed that empty list would mean that Bloop does support it already, but nothing was sent. If the actions filed was None I would assume this is sent by BSP server that doesn't yet support it.

I don't think this matters in the long run, but we can change it if needed.

Reply:
Ah ok ok, you're right. Make sense. 👍🏼